src/types/coordinate_nb.cpp

```diff
@@ -0,0 +1,81 @@
+#include "util/coordinate.hpp"
+
+#include <nanobind/nanobind.h>
+#include <nanobind/operators.h>
+#include <nanobind/stl/string.h>
+#include <nanobind/stl/pair.h>
+
+namespace nb = nanobind;
+using namespace nb::literals;
+
+void init_Coordinate(nb::module_& m) {    
```

Comment:
seems like a ton of coord constructors. does py-osrm/nanobind need all of them to be registered internally? I'd have guessed we only expose one type

Reply:
It looks like Coordinate converts both Float types and UnsafeFloat types to Fixed, ie:
```cpp
inline FixedLatitude toFixed(const FloatLatitude floating)
{
    const auto latitude = static_cast<double>(floating);
    const auto fixed = static_cast<std::int32_t>(std::round(latitude * COORDINATE_PRECISION));
    return FixedLatitude{fixed};
}
```
Would it then make more sense to keep the Fixed constructors, or should I still go with just the one Float constructor?

Reply:
I would use simple Python’s float instead of all these FixedLatitude/Longitude, FloatLatitude/Longitude etc… Just accept float as a parameter and convert to this “safe” type under the hood in glue code… WDYT? Will it be doable with nanobind?

Reply:
It's doable to do away with all the Longitude/Latitude definitions, but one of the 3 types has to be used for the manual constructor for coordinates, as that's what the C++ constructor expects.

```cpp

.def("__init__", [](Coordinate* t, std::pair<double, double> coords) {
    const UnsafeFloatLongitude lon_ = UnsafeFloatLongitude{coords.first};
    const UnsafeFloatLatitude lat_ = UnsafeFloatLatitude{coords.second};

    new (t) Coordinate(lon_, lat_);
 })
```

My main confusion stemmed around if we wanted to treat user input as the Unsafe archetype by default, or Float. Unsafe types use `boost::numeric_cast` to perform an overflow check, while Float types cast it to an `int32_t` (both do the same rounding calculation in the end).

However, FloatCoordinate does not accept UnsafeFloatLongitude/Latitude, though it does accept Float types.

Should I go with Float default for both, and ignore UnsafeFloat?

Reply:
I would just go with FloatLongitude/FloatLatitude and that's it. IMO NodeJs bindings are very good to use as a reference here, just because people are using them for years(i.e. this approach is kind of tested by time)... https://github.com/Project-OSRM/osrm-backend/blob/62e11fd5dbc8d5e888c6e2555505ae684d8a2759/include/nodejs/node_osrm_support.hpp#L413

I quickly looked over codebase and UnsafeFloatLongitude/Latitude is very rarely used for now. There are only 2 usecases: validation of coordinates input in osrm-routed and validation of OSM node coordinates in osrm-extract. Yes, if you would ask me in general, then I would say that Unsafe seems to be much more robust approach - it is a good practice to check external input, BUT since OSRM's codebase is more prepared to use FloatLongitude/FloatLatitude I would go this way for now. After all you will always be able to move to Unsafe version in the future...